contracts/ERC827/proposals/ERC827TokenMockAllowedCallbacks.sol

```diff
@@ -0,0 +1,220 @@
+/* solium-disable security/no-low-level-calls */
+
+pragma solidity ^0.4.24;
+
+import "../../ERC20/StandardToken.sol";
+
+
+/**
+ * @title ERC827, an extension of ERC20 token standard
+ *
+ * @dev Implementation the ERC827, following the ERC20 standard with extra
+ * methods to transfer value and data and execute calls in transfers and
+ * approvals. Uses OpenZeppelin StandardToken.
+ */
+contract ERC827TokenAllowedCallbacks is StandardToken {
+
+  /**
+   * @dev Signatures fo the allowed callback allowed between contracts
+   * Receiver => Sender => Function Signature => Allowed
+   */
+  mapping(address => mapping(address => mapping(bytes4 => bool)))
+    public allowedCallbacks;
+
+  /**
+   * @dev Modifier that check the callback that wants to be executed
+   * If a receiver allows address(0) means that it can be called by anyone
+   * If a receiver allows bytes4(0) means that any function can be called
+   */
+  modifier callbackAllowed(address to, bytes4 functionSignature) {
+    // TO DO: Get first 4 bytes from data instead of using function signature
+    require(isCallbackAllowed(msg.sender, to, functionSignature));
+    _;
+  }
+
+  /**
+   * @dev Allows a callback function to be executed to a receiver contract
+   * @param from address The address that can execute the callback
+   * @param functionSignature bytes4 The signature of the callback function
+   */
+  function allowCallback(address from, bytes4 functionSignature) public {
+    // TO DO: Check that msg.sender is a contract ?
+    allowedCallbacks[msg.sender][from][functionSignature] = true;
+  }
+
+  /**
+   * @dev Remove a callback function to be executed to a receiver contract
+   * @param from address The address that can execute the callback
+   * @param functionSignature bytes4 The signature of the callback function
+   */
+  function removeCallback(address from, bytes4 functionSignature) public {
+    // TO DO: Check that msg.sender is a contract ?
+    allowedCallbacks[msg.sender][from][functionSignature] = false;
+  }
+
+  /**
+   * @dev Addition to ERC20 token methods. It allows to
+   * approve the transfer of value and execute a call with the sent data.
+   * Beware that changing an allowance with this method brings the risk that
+   * someone may use both the old and the new allowance by unfortunate
+   * transaction ordering. One possible solution to mitigate this race condition
+   * is to first reduce the spender's allowance to 0 and set the desired value
+   * afterwards:
+   * https://github.com/ethereum/EIPs/issues/20#issuecomment-263524729
+   * @param _spender The address that will spend the funds.
+   * @param _value The amount of tokens to be spent.
+   * @param _data ABI-encoded contract call to call `_spender` address.
+   * @return true if the call function was executed successfully
+   */
+  function approveAndCall(
+    address _spender,
+    uint256 _value,
+    bytes4 _functionSignature,
```

Comment:
Let's not rely on the sender to provide the signature, but instead take the first 4 bytes of the data.

Here is a quick way to extract them.

```
bytes4 signature;

assembly {
       signature := mload(add(_data, 32))
}
```